rfcs/20230210-sparsity.md

```diff
@@ -0,0 +1,373 @@
+
+As for consideration (1), the most ambitious extension would be to
+allow sparse tensor types for all operations and all operands and
+outputs where regular tensor types are allowed. Such an
+"**open world**" sparsity extension would, in theory, allow
+arbitrary programs written in a language such as JAX or NumPy
+to be converted into an equivalent sparse program by merely
+changing some properties on the tensor (viz. array) types
+of that language. However, generating efficient sparse code
+for such an “open world” system is generally an unsolved problem that
+is still under active research in academia. More realistically,
+in a “**closed world**” sparsity extension, sparse tensor types
+are *restricted to a subset of operations* that can be handled
+efficiently by a downstream “sparse compilation” pass,
+i.e., with some basic guarantees of being able to generate a
+well performing sparse version of the input code.
+
+Although we are open to debate an "open world" extension, this RFC initially
+proposes a "closed world" extension where sparse tensor types can only be
+applied to the following operations (the current set is based on our
+experiences with sparse JAX and MHLO; adding more operations already
+in the initial spec is also open for debate; for the "closed world"
+approach, the set of supported operations should still be easily
+extensible as the need arises).
+
+1. The common arithmetic binary and unary “element-wise” operations:
+   `add`, `subtract`, `multiply`, `negate`
+2. Zero preserving unary math “element-wise” operations:
+   `abs`, `exponential_minus_one`, `log_plus_one`,
+   `sign`, `sine`, `sqrt`, `tanh`, `ceil`, `floor`,
+   `bessel_i1e`, `real`, `imag`
+3. General dot products and reductions:
+   `dot`, `dot_general`, `reduce`
+4. Reshaping operations:
+   `reshape`, `broadcast_in_dim`, `concatenate`, `squeeze`, `transpose`
+5. Conversions (between dense/sparse tensors and between different
+   sparsity encodings): `convert`
+
+As for consideration (2), we could come up with type inference rules
+for all operations that, given the types of the input operands,
+define the types of the result. However, for sparsity, such type
+inference rules would always be heuristic in nature. For example,
+when adding two sparse tensors element-wise, should the result be
+dense (when the two nonzero patterns pretty much cover the complete
+coordinate space) or sparse (when the union of the two nonzero patterns
+still remains sparse). Either heuristic would fail for a certain
+class of sparse tensors. Furthermore, it would introduce difficult
+choices for the other components of sparsity. For example, in the
+following element-wise addition of a row-wise and column-wise
+sparse matrix, it is unclear what ordering should be inferred
+on the output sparse matrix.
+
+      %result = stablehlo.add %matrix1, %matrix2
+        : tensor<10x10xf32, #CSR>,
+          tensor<10x10xf32, #CSC> -> tensor<10x10xf32, #???>

```

Comment:
Ah, I should have read further.

> Furthermore, the IR is not necessarily concerned with the efficiency of any of the expressed operations.

It does seem like supporting code-generation for all possible combinations may place a lot of strain on compiler writers unless there is a default (e.g. like "Always insert a conversion if the mode ordering does not line up.")

Reply:
In my PhD thesis sparse compiler, I had a "panic" default fall back. If none of the efficient techniques worked, it would fall back to simply "looking up" each next element. Very inefficient, but it would at least generate functional code, together with a prompt back to the user that this version would need a better sparsity design.  I feel this approach (or something similar) is slightly better than a priori disallowing sparsity combinations at StableHLO IR level (perhaps some clever researcher finds an efficient way of doing something that was originally deemed to hard ;--)

Reply:
That seems like a reasonable approach. To clarify, I did not intend to restrict the IR, as that seems particularly difficult to do - I was more arguing for disallowing some types of codegen (i.e. "panic" failing to codegen), but you make a good point about it being better to generate inefficient but functional code instead of failing at codegen.